### PR TITLE
mkfs-util: propagate SOURCE_DATE_EPOCH to mcopy

### PR DESCRIPTION
**Pull request by malt3**
If set, mcopy will respect `SOURCE_DATE_EPOCH` and deterministically sets ctime, mtime and atime when creating folders / files in FAT partitions.
Since mcopy is invoked with `execve`, SOURCE_DATE_EPOCH has to be propagated explicitly.

### CODE REVIEW

File: src/shared/mkfs-util.c [hunk 1]
```diff
@@ -159,9 +169,8 @@ static int do_mcopy(const char *node, const char *root) {
         if (r < 0)
                 return r;
         if (r == 0) {
-                /* Avoid failures caused by mismatch in expectations between mkfs.vfat and mcopy by disabling
-                 * the stricter mcopy checks using MTOOLS_SKIP_CHECK. */
-                execve(mcopy, argv, STRV_MAKE("MTOOLS_SKIP_CHECK=1"));
+
+                execve(mcopy, argv, envp);
```
Comment (bluca) [inline src/shared/mkfs-util.c]: should be possible to shorten this to `STRV_MAKE("MTOOLS_SKIP_CHECK=1", source_date_epoch)` as it should handle a nul parameter

Comment (malt3) [reply src/shared/mkfs-util.c]: Good one! One question:
Is my assumption correct that STRV_MAKE will not allocate while strv_new will allocate?

Comment (bluca) [reply src/shared/mkfs-util.c]: yes

File: src/shared/mkfs-util.c [hunk 2]
```diff
@@ -104,9 +104,11 @@ static int mangle_fat_label(const char *s, char **ret) {
 static int do_mcopy(const char *node, const char *root) {
         _cleanup_free_ char *mcopy = NULL;
         _cleanup_strv_free_ char **argv = NULL;
+        _cleanup_strv_free_ char **envp = NULL;
```
Comment (poettering) [inline src/shared/mkfs-util.c]: please merge these two lines

File: src/shared/mkfs-util.c [hunk 3]
```diff
@@ -121,6 +123,14 @@ static int do_mcopy(const char *node, const char *root) {
...
                 return log_error_errno(r, "Failed to determine whether mcopy binary exists: %m");
 
+        /* Avoid failures caused by mismatch in expectations between mkfs.vfat and mcopy by disabling
+        * the stricter mcopy checks using MTOOLS_SKIP_CHECK. */
+        source_date_epoch = strv_find_prefix(environ, "SOURCE_DATE_EPOCH=");
+        if (source_date_epoch) {
+                envp = strv_new(source_date_epoch, "MTOOLS_SKIP_CHECK=1");
+        } else {
+                envp = strv_new("MTOOLS_SKIP_CHECK=1");
+        }
```
Comment (poettering) [inline src/shared/mkfs-util.c]: no {} around single line if blocks, as per coding style

File: src/shared/mkfs-util.c [hunk 4]
```diff
@@ -121,6 +123,14 @@ static int do_mcopy(const char *node, const char *root) {
         if (r < 0)
                 return log_error_errno(r, "Failed to determine whether mcopy binary exists: %m");
 
+        /* Avoid failures caused by mismatch in expectations between mkfs.vfat and mcopy by disabling
+        * the stricter mcopy checks using MTOOLS_SKIP_CHECK. */
+        source_date_epoch = strv_find_prefix(environ, "SOURCE_DATE_EPOCH=");
```
Comment (poettering) [inline src/shared/mkfs-util.c]: hmm? why go via environ[]? why not just getenv(). does pretty much the same

File: src/shared/mkfs-util.c (hunk 3 above)
Comment (poettering) [reply src/shared/mkfs-util.c]: oom check missing

File: src/shared/mkfs-util.c (hunk 4 above)
Comment (malt3) [reply src/shared/mkfs-util.c]: We want to have the full `SOURCE_DATE_EPOCH=1234`

File: src/shared/mkfs-util.c (hunk 1 above)
Comment (poettering) [reply src/shared/mkfs-util.c]: and yeah, strv_new() should not be necessary, just pass this as additional arg to STRV_MAKE()

File: src/shared/mkfs-util.c (hunk 4 above)
Comment (poettering) [reply src/shared/mkfs-util.c]: oh, right, of course

**Comment by poettering** [comment]
lgtm. any reason you still set this to draft?

**Comment by malt3** [comment]
> lgtm. any reason you still set this to draft?

I didn't have time to check if all reproducibility issues are resolved when it comes to FAT.
But let's merge and I'll create a follow up PR if needed.

**Review by DaanDeMeyer — APPROVED** [review]
